quinn/src/endpoint.rs

```diff
@@ -425,6 +429,38 @@ impl State {
     fn drive_recv(&mut self, cx: &mut Context, now: Instant) -> Result<bool, io::Error> {
         self.recv_limiter.start_cycle();
         let mut metas = [RecvMeta::default(); BATCH_SIZE];
+        let rcx = &mut ReceiveContext {
+            cx,
+            now,
+            metas: &mut metas,
+        };
+        let mut poll_res = PollResult::default();
+        if let Some(socket) = &self.abandoned_socket {
+            poll_res = self.poll_socket(&socket.clone(), rcx);
+        };
+        if !poll_res.keep_going {
+            poll_res = self.poll_socket(&self.socket.clone(), rcx);
```

Comment:
Calling `Arc::clone`, and then `Arc::drop`, on every transmit feels a bit ugly. Alternative ideas, in order of preference:
- Group the fields accessed by `poll_socket` into a smaller struct owned by `State`, then move `poll_socket` into a method on that struct and pass in a plain reference to the socket
- Lift `poll_socket` out into a method on `ReceiveContext` instead, adding fields as needed to refer to values owned by `State`

Reply:
Commit 41cb514 implements the first idea.